### PR TITLE
Updates share extension to use new, custom UI

### PR DESCRIPTION
**Pull request by bummytime**
This PR updates the share extension to use a 100% custom UI:

web_share4

photo_share

## Discussion
Let me apologize in advance for the size of this PR. Unfortunately, it is an all or nothing kind of thing since the the share extension is really a mini-app within WPiOS.

There are two main screens:
* The main editing interface: `ShareExtensionEditorViewController.swift` 
* The modular site picker which allows select which site to post to (in subsequent releases we will add additional modules that allow you to further customize the post): `ShareModularViewController.swift`

### Aztec
The editor is powered by Aztec. The new share extension designs called for a subset of the functionality already in `AztecPostViewController.swift`. So, a lot unneeded functionality was stripped out and yielded what you see in `ShareExtensionEditorViewController.swift`. It should be noted that a __lot__ of the dependencies required by Aztec are still needed and were added to the share extension target:

* `MediaAttachment+WordPress.swift`
* `TextList+WordPress.swift`
* `Header+WordPress.swift`
* `FormatBarItemProviders.swift`
* `ImageAttachment+WordPress.swift`
* `OptionsTableView.swift`
* `ImgUploadProcessor.swift`
* `CalypsoProcessorIn.swift`
* `CalypsoProcessorOut.swift`
* `String+RegEx.swift`
* `NSAttributedStringKey+Conversion.swift`

### Modular Site Picker
The modular site picker is pretty bare bones — you pick the site you want to post to and tap “Publish”. Pull-to-refresh was added here to re-fetch the sites in case there is connectivity issues.

### DB, File System, and networking logic
All of the existing file management & persistence logic (to the shared container) as well as site-fetching + media/post-uploading logic pushed in #8179 has remained unchanged. However note that I did consolidate the network functionality into a single spot: `AppExtensionsService.swift`

### Things to note:
* Notifications will be added in a subsequent PR.
* It was an explicit design decision to only allow posting with a “publish” status. If users complain, we will consider adding an option in at a later time.
* Because of occasional memory pressure issues on the iPhone X with `HEIC` images, we are allowing for a maximum of 2 images to be shared.
* There is no retry (yet).
* The site picker will always default to the user’s primary site. We no longer remember the “last used” site as in the previous share extension.
* Since the share extension is a guest within the host app, we are unable to disable rotation. The host app dictates whether the share extension will switch orientations or not. I have tried to make the extension function as best as it can in a compact vertical size class (see iPhone in landscape for my struggles).

Fixes: #8137, #8138, #8139, #8140, #7266
Ref: #8179

### App extension fun
A few things to think about when you are testing:

* Xcode’s debugger prevents the system from suspending the process being debugged.  So, if you run your extension from Xcode, or you attach to its process some time after launch, the process will continue to execute in situations where the system would otherwise have suspended it.  
* Be sure to test on a device (with Xcode disconnected), as well as the simulator. The simulator is much less likely to terminate an extension.
* Try to test on a variety of devices iPhone 5s, iPhone X, __iPads__, etc.
* If you terminate the app by force-quiting from the app switcher, the system will cancel all of the session’s background transfers, and won’t attempt to relaunch the app.

### What to test
Everything sharing. It would probably make sense to focus a lot of time in Photos and Safari, but other apps should work too.

Keep in mind there are different types of sharing:

**Sharing a web page (from Safari):**
In Safari, we are allowed to inject a JS file to extract data from a web page. Currently we are grabbing:
* The URL of the page (document.baseURI)
* A description of the page if that element is provided by the page author
* Any selected text 
* The title of the page (document.title)

**Sharing a URL:**
Sharing a URL should work from many apps on iOS. In this situation we simply create a link and insert it into the content body. It should be noted that even if a URL is shared from Safari, it is very possible we will not be able to extract the other information (title, description, etc). This is simply a limitation of what the extension context provides to us.

**Sharing plain text:**
When you select text from most apps and tap “Share…” in the context menu, this will initiate a plain text share where the only data we can pull from the shared context is the selected text. Here, we simply put that selected text into a blockquote.

**Sharing from WPiOS:**
Yep. In the name of recursion, you can share from WPiOS too.

**Sharing Images:**
Sharing images works as one would expect — all we have to work with is the image data, so that gets embedded into the content body and the title is empty (note there is a maximum of 2).

## Thank you!

Please let me know if you have any questions!

@kwonye would you mind taking point on reviewing this?
@frosty and @SergioEstevao it would be great if I could get your 👀 on this as well!

/cc @diegoreymendez and @jleandroperez for Aztec
/cc @folletto for design

### CODE REVIEW

**Comment by SergioEstevao** [comment]
Some notes after some testing:

- Love that Aztec is being used!

- Couldn't we make the share UI full screen? cover all the screen?

- Maybe this is by design and for the majority of single site users doesn't make a difference, but no search on the site picker?

- Featured images, tags, categories support coming later?

**Comment by kwonye** [comment]
@bummytime the only thing I reviewed was `ShareModularViewController` as it's the only one I was familiar with at all. Looks great! It'll be interesting to see how feasible the tableview delegate methods are once we add the other modules, but I think with only three max it'll work fine.

Great job!

**Comment by SergioEstevao** [comment]
@diegoreymendez I think we really need to take some time and modularise Aztec inside WP to a separate reusable class (maybe even a separate project inside the app). With the current approach we are going to have a lot of duplicated code between the share extension and the main app. Or else when we find bugs or change the behaviour on the main app we will need to remember to track down changes on the share extension...

@bummytime I know we agreed using Aztec like this in the extension, but just want to raise more awareness of future issues.

**Comment by frosty** [comment]
@bummytime This is looking awesome! I'm still working through the code, but here's a few small functional things I noticed while testing:

* The first few times I activated the extension there was no animation for its appearance. But now when I use it it does animate up from the bottom of the screen. Have you seen this before?
* While editing a post, if you tap the List button in the formatting bar, then choose a list style, the size of the editor is incorrect and flows behind the keyboard.
* If I share 2 images, when I post they have no vertical gap between – <p></p>. Would it be best to put the images in separate paragraphs?
* When activated in WPiOS, the extension has a blue nav bar, but in other apps it’s white. Should we standardise this? I guess this might be due to WPiOS's UIAppearance settings?
* I feel like remembering the last used site could be a useful future enhancement. I agree with Sergio that we may want search at some point too (but not for this PR!).

**Comment by folletto** [comment]
> Couldn't we make the share UI full screen? cover all the screen?
> Maybe this is by design and for the majority of single site users doesn't make a difference, but no search on the site picker?
> Featured images, tags, categories support coming later?

All of the above already defined in the design: no full screen, no search (for now), support coming later. :)

File: WordPress/WordPressShareExtension/ShareModularViewController.swift [hunk 1]
```diff
@@ -0,0 +1,552 @@
...
+    @IBOutlet fileprivate var modulesTableView: UITableView!
+
+    /// TableView for site list
+    ///
+    @IBOutlet fileprivate var sitesTableView: UITableView!
+
+    /// Back Bar Button
+    ///
+    fileprivate lazy var backButton: UIBarButtonItem = {
+        let publishTitle = NSLocalizedString("Back", comment: "Back action on share extension site picker screen. Takes the user to the share extension editor screen.")
```
Comment (frosty) [inline WordPress/WordPressShareExtension/ShareModularViewController.swift]: This should probably be named `backTitle` here?

File: WordPress/WordPressShareExtension/ShareModularViewController.swift [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+        let button = UIBarButtonItem(title: publishTitle, style: .plain, target: self, action: #selector(backWasPressed))
+        button.accessibilityIdentifier = "Publish Button"
```
Comment (frosty) [inline WordPress/WordPressShareExtension/ShareModularViewController.swift]: This should be `Back Button` or something similar.

**Comment by bummytime** [comment]
Thanks for all the feedback @frosty. I will make the changes ☝️ you suggested. Regarding this:

> When activated in WPiOS, the extension has a blue nav bar, but in other apps it’s white. Should we standardise this? I guess this might be due to WPiOS's UIAppearance settings?

That is exactly why that is happening. I left it alone because I assumed this made sense when viewing the share extension within the context of WPiOS.

**Comment by bummytime** [comment]
@frosty ready for another look. Three of the comments were not addressed:

> The first few times I activated the extension there was no animation for its appearance. But now when I use it it does animate up from the bottom of the screen. Have you seen this before?

Yeah, unfortunately I have seen this...and it seems to be a little random. It's like `UIViewControllerAnimatedTransitioning`'s `animateTransition` call is unable to fire (perhaps when the `ShareExtractor` is loading a lot of data out of the share context). I noticed this happens a lot more when sharing images vs. text. FWIW, I did end up moving `loadContent()` into `ShareExtensionEditorViewController.viewDidLoad()` to try and buy us a few extra processor cycles. Not sure it really helped though.

> If I share 2 images, when I post they have no vertical gap between – `<p></p>`. Would it be best to put the images in separate paragraphs?

The code for inserting the images is [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/a20e332c5ed4955072ab21c46b98217c97e88792/WordPress/WordPressShareExtension/ShareExtensionEditorViewController.swift#L1202-L1206) and [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/a20e332c5ed4955072ab21c46b98217c97e88792/WordPress/WordPressShareExtension/ShareExtensionEditorViewController.swift#L968-L975). I am wondering if @SergioEstevao would have some insight on spacing/padding best practices when adding multiple images into Aztec.

> While editing a post, if you tap the List button in the formatting bar, then choose a list style, the size of the editor is incorrect and flows behind the keyboard.

Still working on this one I should have something tomorrow morning.

😄

**Comment by bummytime** [comment]
@folletto in the name of not failing silently, I have a catch-all alert that could display after `Publish` is tapped. Admittedly, the chances of this are rare — it is more or less a parachute in case an error we can't predict yet, happens.

Sooooo, the dialog looks like this:

error

Opinions?

File: WordPress/WordPressShareExtension/ShareModularViewController.swift [hunk 3]
```diff
@@ -471,25 +473,54 @@ fileprivate extension ShareModularViewController {
...
+            }, onFailure: {
+                let error = self.createErrorWithDescription("Failed to save and upload post with no media.")
+                self.tracks.trackExtensionError(error)
+                self.showAlert()
             })
         }
     }
+
+    func showAlert() {
+        let title = NSLocalizedString("Error Occured While Sharing", comment: "Share extension error dialog title - displayed when an error occurs while sharing.")
```
Comment (frosty) [inline WordPress/WordPressShareExtension/ShareModularViewController.swift]: Typo: "Occured" -> "Occurred"

**Comment by folletto** [comment]
> Admittedly, the chances of this are rare — it is more or less a parachute in case an error we can't predict yet, happens.

It's good. As far I understand being a catch-all means we can't be specific about what happened, right?

I've tweaked the text a bit:

> **Sharing Error**
> Whoops, something went wrong while sharing.
> You can try again, maybe it was a glitch.
> [ Try again ]
> [ Nevermind ]

**Comment by bummytime** [comment]
@folletto

> As far I understand being a catch-all means we can't be specific about what happened, right?

That is correct. 😄  

Updated the dialog (I forced the error to happen here):

share_error

**Comment by folletto** [comment]
Ace 👌

**Comment by bummytime** [comment]
@frosty keyboard glitch with the options VC should be fixed. Ready for another 👀  

Thanks❗️❗️❗️

**Comment by SergioEstevao** [comment]
```The code for inserting the images is here and here. I am wondering if @SergioEstevao would have some insight on spacing/padding best practices when adding multiple images into Aztec.```

In Aztec we insert images without a paragraph break, but it's something I think we should change because it looks strange on the end post and it also makes the cursor caret in Aztec behave in erratic ways.

**Comment by bummytime** [comment]
@kwonye were you able to do any additional testing on this?

**Comment by bummytime** [comment]
Thanks for the thorough review @frosty!!!!

**Comment by bummytime** [comment]
Thanks for reviewing this HUGE PR @frosty @SergioEstevao @kwonye. Much appreciated! 💥

**Comment by jleandroperez** [comment]
@bummytime i'm late to the party, just stepping by to say **I LOVE WHAT YOU DID!!!!**

simulator screen shot - iphone se - 2018-01-29 at 14 15 41

**Comment by bummytime** [comment]
Haha. Thanks @jleandroperez!

**Pull request merged**